### PR TITLE
[PhpUnitBridge] deprecate the testLegacy test name prefix

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #18755
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 1]
```diff
@@ -197,6 +197,11 @@ public function endTest(\PHPUnit_Framework_Test $test, $time)
                 DnsMock::withMockedHosts(array());
             }
         }
+
+        if ($test instanceof \PHPUnit_Framework_TestCase && 0 === strpos($test->getName(), 'testLegacy') && empty($test->getTestResultObject()->warningCount())) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: when the annotation is in place, we should not trigger the warning. What about the other cases?
From DeprecationErrorHandler:
```php
                } elseif (0 === strpos($method, 'testLegacy')
                    || 0 === strpos($method, 'provideLegacy')
                    || 0 === strpos($method, 'getLegacy')
                    || strpos($class, '\Legacy')
```

Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: Adding the warning based on class/method names does work. I am looking into how to do that (if possible) based on the data transformer.

**Comment by xabbuh** [comment]
Status: Needs work

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 2]
```diff
@@ -170,6 +172,13 @@ public function startTest(\PHPUnit_Framework_Test $test)
         }
     }
 
+    public function addWarning(\PHPUnit_Framework_Test $test, \PHPUnit_Framework_Warning $e, $time)
+    {
+        if ($test instanceof \PHPUnit_Framework_TestCase) {
```
Comment (OskarStark) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: why do we need this check here, we already typehint the parameter

Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: The type hint is for `\PHPUnit_Framework_Test`, but not `\PHPUnit_Framework_TestCase`.

Comment (OskarStark) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: damn, you are right 👍

**Comment by xabbuh** [comment]
The `provideLegacy` and `getLegacy` prefixes are not used to make the test belong to the legacy group, but can be used if your data provider triggers deprecations itself. This is something that needs to be fixed in the documentation (see symfony/symfony-docs#7312 for the bug report). Thus, this PR is finished and ready to be reviewed.

Status: Needs review

**Comment by xabbuh** [comment]
After talking to @nicolas-grekas the warnings will now only be triggered if the `@group` annotation was not used. So you can get rid of the warnings simply by adding the group without the need to rename your test methods.

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 3]
```diff
@@ -188,15 +201,21 @@ public function endTest(\PHPUnit_Framework_Test $test, $time)
...
             if (in_array('time-sensitive', $groups, true)) {
                 ClockMock::withClockMock(false);
             }
             if (in_array('dns-sensitive', $groups, true)) {
                 DnsMock::withMockedHosts(array());
             }
         }
+
+        if ($test instanceof \PHPUnit_Framework_TestCase && 0 === strpos($test->getName(), 'testLegacy') && !isset($this->testsWithWarnings[$test->getName()]) && !in_array('legacy', $groups, true)) {
+            $test->getTestResultObject()->addWarning($test, new \PHPUnit_Framework_Warning('Using the testLegacy prefix to mark tests as legacy is deprecated since version 3.3 and will be removed in 4.0. Use the "@group legacy" notation instead to add the test to the legacy group.'), $time);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: quotes around "testLegacy"

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 4]
```diff
@@ -188,15 +201,21 @@ public function endTest(\PHPUnit_Framework_Test $test, $time)
...
+        }
+
+        if ($test instanceof \PHPUnit_Framework_TestCase && strpos($className, '\Legacy') && !isset($this->testsWithWarnings[$test->getName()]) && !in_array('legacy', $classGroups, true)) {
+            $test->getTestResultObject()->addWarning($test, new \PHPUnit_Framework_Warning('Using the Legacy prefix to mark all tests of a class as legacy is deprecated since version 3.3 and will be removed in 4.0. Use the "@group legacy" notation instead to add the test to the legacy group.'), $time);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: quotes around "Legacy"

**Review by nicolas-grekas — COMMENTED** [review]
:+1: with minor comments

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: quotes added in both places

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**